File: src/ReadStorm.Infrastructure/Services/FastSourceHealthCheckUseCase.cs
using System.Text;
using ReadStorm.Application.Abstractions;
using ReadStorm.Domain.Models;

namespace ReadStorm.Infrastructure.Services;

/// <summary>
/// 并发关键词探活：对每个书源发起一次搜索请求（GET/POST），
/// 单源 3 秒超时，全量并发一次完成。
/// </summary>
public sealed class FastSourceHealthCheckUseCase : ISourceHealthCheckUseCase
{
    private static readonly TimeSpan PerSourceTimeout = TimeSpan.FromSeconds(3);

    private readonly HttpClient _httpClient;
    private readonly IReadOnlyList<string> _ruleDirectories;

    public FastSourceHealthCheckUseCase(HttpClient? httpClient = null, IReadOnlyList<string>? ruleDirectories = null)
    {
        _httpClient = httpClient ?? RuleHttpHelper.CreateHealthCheckHttpClient(PerSourceTimeout);
        _ruleDirectories = ruleDirectories ?? RulePathResolver.ResolveAllRuleDirectories();
    }

    public async Task<IReadOnlyList<SourceHealthResult>> CheckAllAsync(
        IReadOnlyList<BookSourceRule> sources,
        CancellationToken cancellationToken = default)
    {
        // 全量并发：每个书源独立超时，总耗时 ≈ 最慢的一个。
        var tasks = sources
            .Where(s => s.Id > 0 && !string.IsNullOrWhiteSpace(s.Url))
            .Select(s => PingOneAsync(s.Id, s.Url, cancellationToken))
            .ToList();

        var results = await Task.WhenAll(tasks);
        return results;
    }

    private async Task<SourceHealthResult> PingOneAsync(
        int sourceId, string url, CancellationToken cancellationToken)
    {
        try
        {
            using var cts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
            cts.CancelAfter(PerSourceTimeout);

            var request = await BuildSearchProbeRequestAsync(sourceId, url, "测试", cts.Token);
            if (request is null)
            {
                return new SourceHealthResult(sourceId, false);
            }

            using (request)
            {
                using var response = await _httpClient.SendAsync(
                    request, HttpCompletionOption.ResponseHeadersRead, cts.Token);

                var code = (int)response.StatusCode;
                return new SourceHealthResult(sourceId, code is >= 200 and < 400);
            }
        }
        catch
        {
            return new SourceHealthResult(sourceId, false);
        }
    }

    private async Task<HttpRequestMessage?> BuildSearchProbeRequestAsync(
        int sourceId,
        string fallbackUrl,
        string keyword,
        CancellationToken cancellationToken)
    {
        var rule = await LoadRuleAsync(sourceId, cancellationToken);
        var search = rule?.Search;
        if (search is null || string.IsNullOrWhiteSpace(search.Url))
        {
            // 没有搜索规则时回退到首页 GET 探活（不再使用 HEAD）
            return new HttpRequestMessage(HttpMethod.Get, fallbackUrl);
        }

        var escapedKeyword = Uri.EscapeDataString(keyword);
        var resolvedUrl = search.Url.Contains("%s", StringComparison.Ordinal)
            ? search.Url.Replace("%s", escapedKeyword, StringComparison.Ordinal)
            : search.Url;

        var isPost = string.Equals(search.Method, "post", StringComparison.OrdinalIgnoreCase);
        var request = new HttpRequestMessage(isPost ? HttpMethod.Post : HttpMethod.Get, resolvedUrl);

        if (!string.IsNullOrWhiteSpace(search.Cookies))
        {
            request.Headers.TryAddWithoutValidation("Cookie", search.Cookies);
        }

        if (isPost)
        {
            var payload = BuildFormBody(search.Data, keyword);
            request.Content = new StringContent(payload, Encoding.UTF8, "application/x-www-form-urlencoded");
        }

        return request;
    }

    private Task<RuleFileDto?> LoadRuleAsync(int sourceId, CancellationToken cancellationToken)
        => RuleFileLoader.LoadRuleAsync(_ruleDirectories, sourceId, cancellationToken);

    private static string BuildFormBody(string? rawData, string keyword)
    {
        if (string.IsNullOrWhiteSpace(rawData))
        {
            return $"searchkey={Uri.EscapeDataString(keyword)}";
        }

        var text = rawData.Trim();
        if (text.StartsWith("{", StringComparison.Ordinal) && text.EndsWith("}", StringComparison.Ordinal) && text.Length >= 2)
        {
            text = text[1..^1];
        }

        var parts = text.Split(',', StringSplitOptions.RemoveEmptyEntries | StringSplitOptions.TrimEntries);
        var pairs = new List<string>(parts.Length);

        foreach (var part in parts)
        {
            var idx = part.IndexOf(':');
            if (idx <= 0)
            {
                continue;
            }

            var key = part[..idx].Trim().Trim('\'', '"');
            var value = part[(idx + 1)..].Trim().Trim('\'', '"');
            if (string.IsNullOrWhiteSpace(key))
            {
                continue;
            }

            value = value.Replace("%s", keyword, StringComparison.Ordinal);
            pairs.Add($"{Uri.EscapeDataString(key)}={Uri.EscapeDataString(value)}");
        }

        if (pairs.Count == 0)
        {
            return $"searchkey={Uri.EscapeDataString(keyword)}";
        }

        return string.Join("&", pairs);
    }
}

File: pureopus/app/src/main/java/com/readstorm/app/infrastructure/services/FastSourceHealthCheckUseCase.kt
package com.readstorm.app.infrastructure.services

import android.content.Context
import com.readstorm.app.application.abstractions.ISourceHealthCheckUseCase
import com.readstorm.app.application.abstractions.SourceHealthResult
import com.readstorm.app.domain.models.BookSourceRule
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.async
import kotlinx.coroutines.awaitAll
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.withContext
import okhttp3.FormBody
import okhttp3.OkHttpClient
import okhttp3.Request
import java.net.URLEncoder

class FastSourceHealthCheckUseCase(private val context: Context) : ISourceHealthCheckUseCase {

    private val httpClient: OkHttpClient = RuleHttpHelper.createHealthCheckHttpClient()

    override suspend fun checkAll(sources: List<BookSourceRule>): List<SourceHealthResult> =
        coroutineScope {
            sources
                .filter { it.id > 0 && it.url.isNotBlank() }
                .map { source -> async { pingOne(source.id, source.url) } }
                .awaitAll()
        }

    private suspend fun pingOne(sourceId: Int, url: String): SourceHealthResult =
        withContext(Dispatchers.IO) {
            try {
                val request = buildSearchProbeRequest(sourceId, url, "测试")
                    ?: return@withContext SourceHealthResult(sourceId, false)
                val response = httpClient.newCall(request).execute()
                response.use { resp ->
                    SourceHealthResult(sourceId, resp.code in 200..399)
                }
            } catch (_: Exception) {
                SourceHealthResult(sourceId, false)
            }
        }

    private fun buildSearchProbeRequest(sourceId: Int, fallbackUrl: String, keyword: String): Request? {
        val rule = try {
            RuleFileLoader.loadRule(context, sourceId)
        } catch (_: Exception) {
            null
        }
        val search = rule?.search
        if (search == null || search.url.isBlank()) {
            return Request.Builder().url(fallbackUrl).get().build()
        }

        val escapedKeyword = URLEncoder.encode(keyword, "UTF-8")
        val resolvedUrl = if (search.url.contains("%s")) {
            search.url.replace("%s", escapedKeyword)
        } else search.url

        val isPost = search.method.equals("post", ignoreCase = true)
        val builder = Request.Builder().url(resolvedUrl)

        if (search.cookies.isNotBlank()) {
            builder.header("Cookie", search.cookies)
        }

        if (isPost) {
            val payload = buildFormBody(search.data, keyword)
            builder.post(payload)
        }

        return builder.build()
    }

    private fun buildFormBody(rawData: String?, keyword: String): FormBody {
        val fb = FormBody.Builder()
        if (rawData.isNullOrBlank()) {
            fb.add("searchkey", keyword)
            return fb.build()
        }

        var text = rawData.trim()
        if (text.startsWith("{") && text.endsWith("}") && text.length >= 2) {
            text = text.substring(1, text.length - 1)
        }

        val parts = text.split(',').map { it.trim() }.filter { it.isNotBlank() }
        var hasPairs = false
        for (part in parts) {
            val idx = part.indexOf(':')
            if (idx <= 0) continue
            val key = part.substring(0, idx).trim().trim('\'', '"')
            var value = part.substring(idx + 1).trim().trim('\'', '"')
            if (key.isBlank()) continue
            value = value.replace("%s", keyword)
            fb.add(key, value)
            hasPairs = true
        }

        if (!hasPairs) {
            fb.add("searchkey", keyword)
        }

        return fb.build()
    }
}
